gh-93343: Expand warning filter examples

daniel-shimon opened the pull request:
Add examples of warning filters and the difference between programatic and environmental filters.


<!-- gh-issue-number: gh-93343 -->
* Issue: gh-93343
<!-- /gh-issue-number -->


<!-- readthedocs-preview cpython-previews start -->
----
:books: Documentation preview :books:: https://cpython-previews--106618.org.readthedocs.build/

<!-- readthedocs-preview cpython-previews end -->

ezio-melotti commented on Doc/library/warnings.rst:
```diff
@@ -214,6 +214,18 @@ Some examples::
 
+Note that :func:`filterwarnings` filters have subtle differences from :option:`-W` and :envvar:`PYTHONWARNINGS`::
+
+   filterwarnings("ignore", message=".*generic", module=r"yourmodule\.submodule")
```
```suggestion
   filterwarnings("ignore", message="\.*generic", module="yourmodule\.submodule")
```
Shouldn't this be escaped as well?  The `r` here is not necessary, and its usage should be consistent between the two args (assuming they are both regexes).

daniel-shimon replied:
No since the first arg wants to capture strings that contain 'generic' so that the '.' catches everything, while the first arg want to capture 'yourmodule.submodule' specifically, meaning that the '.' actually captures dot and should be escaped

erlend-aasland replied:
Well, it sounds to me you should put that explanation into the docs, @daniel-shimon; if it is unclear for a reviewer, it is not going to be clear for the average docs reader ;)

ezio-melotti commented on Doc/library/warnings.rst:
```diff
@@ -214,6 +214,18 @@ Some examples::
+
+   filterwarnings("ignore", message=".*generic", module=r"yourmodule\.submodule")
+      # Ignore warnings in "yourmodule.submodule" which contain "generic"
+   filterwarnings("ignore", module="yourmodule.*")
```
```suggestion
   filterwarnings("ignore", module="yourmodule\.*")
```
I think this should be escaped too.

erlend-aasland replied:
Also, instead of using comments, would it be clearer (formatting wise) to use multiple code blocks? (Goes for the change below as well)

daniel-shimon replied:
No because here we actually want to catch all characters after 'yourmodule' and not only dot

erlend-aasland replied:
Ditto: https://github.com/python/cpython/pull/106618/files#r1261710032

ezio-melotti commented on Doc/library/warnings.rst:
```diff
@@ -214,6 +214,18 @@ Some examples::
+
+   -W "ignore:generic::yourmodule.submodule:"
+      # Ignore warnings in "yourmodule.submodule" which START with "generic"
+      # Also note that the '.' in the module does not need to be escaped
```
This should also clarify *why* it doesn't need to be escaped.

ezio-melotti commented on Misc/NEWS.d/next/Documentation/2023-07-11-08-46-13.gh-issue-93343.fw_eyw.rst:
```diff
@@ -0,0 +1,2 @@
+Add examples of warning filters and the difference between programatic and
+environmental filters.
```
```suggestion
Add examples of warning filters and the difference between programmatic and
environmental filters.
```

ezio-melotti commented on Doc/library/warnings.rst:
```diff
@@ -214,6 +214,18 @@ Some examples::
    ignore,default:::mymodule    # Only report warnings triggered by "mymodule"
    error:::mymodule             # Convert warnings to errors in "mymodule"
 
+Note that :func:`filterwarnings` filters have subtle differences from :option:`-W` and :envvar:`PYTHONWARNINGS`::
```
```suggestion
Note that :func:`filterwarnings` filters have subtle differences
from :option:`-W` and :envvar:`PYTHONWARNINGS`::
```

ezio-melotti reviewed (commented):
Thanks for the PR!
This is a step in the right direction, but there are more subtleties in https://github.com/pytest-dev/pytest/discussions/8759#discussioncomment-2818984 that are not described here.

In particular:
* Where/when can I uses regexes vs plain text?  This applies to both different types of filters (`filterwarnings()`/`-W`/`PYTHONWARNINGS`) and different fields within a filter (message/package/module).
* How to match part of a message with the different types of filters (beginning of the message vs substring).
* How to filter messages coming from specific packages/modules/submodules with the different types of filters.

I think this would be better covered in a new section of the `warnings` docs -- either a generic `Examples` section that includes these alongside with comments and explanations, or a more specific section that focuses on the differences between the different warnings types.  A mini-howto might also be an option, but we can start with a single section and expand it later.

In addition, since you would have to test different filters to ensure that the behavior you are documenting is correct, it might be worth adding some of these examples to the `warnings` tests in `Lib/test`.

daniel-shimon committed:
gh-93343: Expand warning filter examples

    Add examples of warning filters and the difference between programatic and environmental filters.

daniel-shimon committed:
gh-93343: Add warning filters tests

hugovk reviewed (changes requested):
Ping :)

daniel-shimon committed:
Clarify escaped dots in regexes

daniel-shimon committed:
Merge branch 'main' into gh-93343-extra-warning-examples

picnixz commented on Doc/library/warnings.rst:
```diff
@@ -229,6 +229,39 @@ Some examples::
+
+   filterwarnings("ignore", message=".*generic", module=r"yourmodule\.submodule")
+      # Ignore warnings in "yourmodule.submodule" which contain "generic".
+      # Note that the '.' in 'message' marks any character and in 'module' it is escaped,
+      #   in order to match a literal dot character.
+   filterwarnings("ignore", message="generic", module=r"yourmodule\.submodule")
```
I think those examples should be separated by a new line for clarity purposes.

picnixz commented on Lib/test/test_warnings/__init__.py:
```diff
@@ -1352,6 +1364,18 @@ def test_single_warning(self):
 
+    def test_string_literals(self):
+        # Ensure message/module are treated as string literals
+        rc, stdout, stderr = assert_python_ok("-c",
```
If you use `rc`, check its value as well. Otherwise use `_, stdout, stderr = ...`

daniel-shimon reviewed (commented):
Fix CR Comments

daniel-shimon committed:
Fix CR Comments
